common/pkg/capi/clustertopology/handlers/lifecycle/parallel.go

```diff
@@ -0,0 +1,292 @@
+// Copyright 2024 Nutanix. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package lifecycle
+
+import (
+	"context"
+	"reflect"
+	"strings"
+	"sync"
+
+	"github.com/samber/lo"
+	runtimehooksv1 "sigs.k8s.io/cluster-api/exp/runtime/hooks/api/v1alpha1"
+	"sigs.k8s.io/cluster-api/util"
+)
+
+func runHooksInParallel[T runtimehooksv1.RequestObject, U runtimehooksv1.ResponseObject](
+	ctx context.Context, hookFuncs []func(context.Context, T, U), request T, response U,
+) {
+	responseChan := make(chan U, len(hookFuncs))
+
+	responsePtrType := reflect.TypeFor[U]().Elem()
+
+	var wg sync.WaitGroup
+	wg.Add(len(hookFuncs))
+
+	for _, f := range hookFuncs {
+		hookResponse := reflect.New(responsePtrType).Interface().(U)
+		go func() {
+			f(ctx, request, hookResponse)
+			responseChan <- hookResponse
+			wg.Done()
+		}()
+	}
+
+	// Important final goroutine to wait until WaitGroup is done
```

Comment:
Just checking, each hook runs independently and allows to either succeed, error or timeout?

Reply:
Yes, I've added comments throughout this function now.